src/goto-programs/goto_program.cpp

```diff
@@ -709,13 +710,44 @@ void goto_programt::instructiont::validate(
         const auto &goto_id = goto_symbol_expr.get_identifier();
 
         if(!ns.lookup(goto_id, table_symbol))
+        {
+          bool symbol_expr_type_matches_symbol_table =
+            base_type_eq(goto_symbol_expr.type(), table_symbol->type, ns);
+
+          if(
+            !symbol_expr_type_matches_symbol_table &&
+            table_symbol->type.id() == ID_code)
+          {
+            // Return removal sets the return type of a function symbol table
+            // entry to 'void', but some callsites still expect the original
+            // type (e.g. if a function is passed as a parameter)
+
+            // Check whether return removal has happend
+            const irep_idt return_value_symbol_name =
+              id2string(goto_id) + RETURN_VALUE_SUFFIX;
+            const symbolt *return_value_symbol;
+
+            if(!ns.lookup(return_value_symbol_name, return_value_symbol))
+            {
+              // Return removal has been applied. Temporarily rebuild the full
+              // type of the function and use that full type to check
+              // consistency with the goto program
+              auto full_table_symbol_type = to_code_type(table_symbol->type);
+              full_table_symbol_type.return_type() = return_value_symbol->type;
```

Comment:
This can be simplified by using `original_return_type` (from `remove_returns.{h,cpp}`).

Reply:
Thanks - I did actually look for a utility function to use here, but somehow managed to miss that. I much appreciate the pointer.